presto-pinot/src/main/java/com/facebook/presto/pinot/PinotPageSourceProvider.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.pinot;
+
+import com.facebook.presto.spi.ColumnHandle;
+import com.facebook.presto.spi.ConnectorPageSource;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.ConnectorSplit;
+import com.facebook.presto.spi.connector.ConnectorPageSourceProvider;
+import com.facebook.presto.spi.connector.ConnectorTransactionHandle;
+import com.google.common.collect.ImmutableList;
+
+import javax.inject.Inject;
+
+import java.util.List;
+
+import static com.facebook.presto.pinot.PinotUtils.checkType;
+import static java.util.Objects.requireNonNull;
+
+public class PinotPageSourceProvider
+        implements ConnectorPageSourceProvider
+{
+    private final PinotConfig pinotConfig;
+    private final PinotScatterGatherQueryClient pinotQueryClient;
+
+    @Inject
+    public PinotPageSourceProvider(PinotConfig pinotConfig, PinotScatterGatherQueryClient pinotQueryClient)
+    {
+        this.pinotConfig = requireNonNull(pinotConfig, "pinotConfig is null");
+        this.pinotQueryClient = requireNonNull(pinotQueryClient, "pinotQueryClient is null");
+    }
+
+    @Override
+    public ConnectorPageSource createPageSource(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorSplit split, List<ColumnHandle> columns)
+    {
+        requireNonNull(split, "partitionChunk is null");
+        PinotSplit pinotSplit = checkType(split, PinotSplit.class, "split");
+
+        ImmutableList.Builder<PinotColumnHandle> handlesBuilder = ImmutableList.builder();
+        if (columns.isEmpty()) {
```

Comment:
It doesn't seems to be responsibility of this class. If 0 columns is requested by the engine - the output page should contain zero columns. Otherwise the contract will be broken.